internal/proxy/directors.go

```diff
@@ -185,51 +184,47 @@ type HTTPHeader struct {
 	Value string
 }
 type HTTPHeaders []HTTPHeader
-type NetHeaders struct {
-	IPNet   net.IPNet
-	Headers HTTPHeaders
+
+type Net struct {
+	net.IPNet
+}
+
+func (n Net) Network() net.IPNet { return n.IPNet }
+
+type DirectorSetHeadersByIP struct {
+	allHeaders []string
+	cidranger.Ranger[HTTPHeaders]
 }
-type DirectorSetHeadersByIP []NetHeaders
 
 func NewDirectorSetHeadersByIP(m map[string]HTTPHeaders) (DirectorSetHeadersByIP, error) {
-	res := make(DirectorSetHeadersByIP, 0, len(m))
+	allHeaders := make([]string, 0, 100)
+
+	ranger := cidranger.NewPCTrieRanger[HTTPHeaders]()
 	for k, v := range m {
 		_, subnet, err := net.ParseCIDR(k)
 		if err != nil {
-			return nil, fmt.Errorf("can't parse CIDR: %v %w", k, err)
+			return DirectorSetHeadersByIP{}, fmt.Errorf("can't parse CIDR: %v %w", k, err)
 		}
 
-		res = append(res, NetHeaders{
-			IPNet:   *subnet,
-			Headers: v,
-		})
-	}
-	sortByIPNet(res)
-	return res, nil
-}
-
-func sortByIPNet(d DirectorSetHeadersByIP) {
-	sort.Slice(d, func(i, j int) bool {
-		left, right := d[i], d[j]
+		err = ranger.Insert(&Net{IPNet: *subnet}, v)
+		if err != nil {
+			return DirectorSetHeadersByIP{}, fmt.Errorf("can't insert into cidranger %w", err)
+		}
 
-		maskOnes := func(m net.IPMask) int {
-			ones, _ := m.Size()
-			return ones
+		for _, header := range v {
+			allHeaders = append(allHeaders, header.Name)
```

Comment:
At this point header.Name must be normalized by http.CanonicalHeaderKey - for available check key in map later.